flock_agent/daemon/daemon.py

```diff
@@ -357,3 +361,96 @@ def log(self, request, response, time):
+                )
+                release = r.json()
+                latest_version = release["tag_name"].lstrip("v")
+                self.c.log(
+                    "Daemon",
+                    "autoupdate_loop",
+                    f"installed version: {self.c.version}, latest version: {latest_version}",
+                )
+                if latest_version <= self.c.version:
+                    await asyncio.sleep(update_check_delay)
+                    continue
+
+                # Find the pkg asset
+                url = None
+                filename = None
+                for asset in release["assets"]:
+                    if asset["name"].endswith(".pkg"):
+                        url = asset["browser_download_url"]
+                        filename = asset["name"]
+                        break
+
+                if not url:
+                    self.c.log("Daemon", "autoupdate_loop", "could not find .pkg file")
+                    await asyncio.sleep(update_check_delay)
+                    continue
+
+                # Download the update
+                self.c.log("Daemon", "autoupdate_loop", f"downloading {url}")
+                r = requests.get(url)
+
+                os.makedirs(os.path.join(self.lib_dir, "updates"), exist_ok=True)
+                download_filename = os.path.join(self.lib_dir, "updates", filename)
+                with open(download_filename, "wb") as f:
+                    f.write(r.content)
+
+                self.c.log(
+                    "Daemon",
+                    "autoupdate_loop",
+                    f"download complete: {download_filename}",
+                )
+
+                # Verify that it's codesigned
+                p = subprocess.run(
+                    ["/usr/sbin/pkgutil", "--check-signature", download_filename],
+                    stdout=subprocess.PIPE,
+                )
+                if (
+                    p.returncode != 0
+                    or "Status: signed by a developer certificate issued by Apple for distribution"
+                    not in p.stdout.decode()
+                    or "Developer ID Installer: FIRST LOOK PRODUCTIONS, INC. ("
+                    not in p.stdout.decode()
+                ):
+                    self.c.log(
+                        "Daemon",
+                        "autoupdate_loop",
+                        f"codesign verification failed: {p.stdout.decode()}",
+                    )
+                    await asyncio.sleep(update_check_delay)
+                    continue
+
+                # Install the update
+                self.c.log(
+                    "Daemon",
+                    "autoupdate_loop",
+                    "launching installer background process and quitting daemon",
+                )
+                subprocess.Popen(
+                    ["/usr/sbin/installer", "-pkg", download_filename, "-target", "/"]
```

Comment:
Any sense of the states of broken this could leave the package in? And if subprocess.Popen returns anything other than 0, do we still want to restart it?

Reply:
Here's why this is tricky. [This](https://github.com/firstlookmedia/flock-agent/blob/master/install/macos/build_pkg.py) is the script that builds the pkg file. In order to make upgrading smoother, it includes [these](https://github.com/firstlookmedia/flock-agent/tree/master/install/macos/packaging/scripts) `preinstall` and `postinstall` scripts.

So when the daemon runs `/usr/sbin/installer` to install the new pkg, before installing it will run `preinstall`, which includes running this:

```
/bin/launchctl unload /Library/LaunchDaemons/media.firstlook.flock-agentd.plist
```

This unloads the daemon's launchd plist, which stops the daemon. So, if everything is going well, the daemon will get killed before the subprocess.Popen process is done running anyway.

Then, after unpacking the package, the `postinstall` script runs this:

```
/Applications/Flock.app/Contents/Resources/share/autostart/macos/enable_daemon
```

The source for `enable_daemon` is [here](https://github.com/firstlookmedia/flock-agent/blob/master/share/autostart/macos/enable_daemon), and that reloads the daemon launchd plist, which will start a process of the new version of the daemon.

This is the reason I quit immediately. Because the installer is about to make it quit anyway, and then after installing it's going to run the new version.